### PR TITLE
Migration launch

### PR DESCRIPTION
**Pull request by Drab-Hounet**

### CODE REVIEW

File: hammr/utils/migration_utils.py [hunk 1]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
...
 
     return table
+
+
+def check_account_from_platform(accounts, credAccountName, targetPlatformName):
+    if accounts is None or len(accounts) == 0:
+        printer.out(
+            "No accounts available on the plateform.\n You can use the command 'hammr account create' to create an account.",
+            printer.ERROR)
+        return
```
Comment (segalaj) [inline hammr/utils/migration_utils.py]: shouldn't we raise an exception to be coherent with other functions?

File: hammr/utils/migration_utils.py [hunk 2]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
...
+    else:
+        for account in accounts:
+            if account.name == credAccountName:
+                if account.targetPlatform.name == targetPlatformName:
+                    return account
+        return
```
Comment (segalaj) [inline hammr/utils/migration_utils.py]: shouldn't we raise an exception to be coherent with other functions?

File: hammr/commands/migration/migration.py [hunk 3]
```diff
@@ -48,8 +60,170 @@ def do_list(self, args):
...
+
+        dlUtils = download_utils.Download(download_url, local_uforge_migration_path,
+                                          not self.api.getDisableSslCertificateValidation())
+        try:
+            dlUtils.start()
+        except Exception:
+            return
+        return local_uforge_migration_path
+
+    def deploy_and_launch_agent(self, uforge_login, uforge_password, migrationConfig, file_src_path, uforge_url):
```
Comment (segalaj) [inline hammr/commands/migration/migration.py]: this method contains duplicated code from scan.py, it should be extracted and reuse

File: hammr/commands/migration/migration.py [hunk 4]
```diff
@@ -48,8 +60,170 @@ def do_list(self, args):
...
+                doArgs = doParser.parse_args(shlex.split(args))
+            except SystemExit:
+                return 2
+
+            migrationConfig = migration_utils.check_migration_configuration(doArgs.file)
+            targetFormat, targetPlatform = migration_utils.check_target(self.api, self.login, migrationConfig)
+            myPublishImage = migration_utils.check_publish_image(migrationConfig, targetFormat, doArgs.file)
+            myCredAccount = migration_utils.check_credAccount(self.api, self.login, migrationConfig, targetPlatform)
+
+            printer.out("file: " + doArgs.file + " is validated", "OK")
```
Comment (segalaj) [inline hammr/commands/migration/migration.py]: you should use printer.OK instead of hard coded "OK"

File: hammr/commands/migration/migration.py [hunk 5]
```diff
@@ -48,8 +60,170 @@ def do_list(self, args):
...
+
+            myMigration = self.createMigration(migrationConfig, targetFormat, myCredAccount, myPublishImage)
+
+            # call UForge API
+            self.api.Users(self.login).Migrations.Create(body=myMigration, element_name="ns1:migration")
+
+            # download migration binary
+            local_uforge_migration_path = self.download_migration_binary()
```
Comment (segalaj) [inline hammr/commands/migration/migration.py]: style are mixed for variable names: myVar and my_var

File: hammr/commands/migration/migration.py [hunk 6]
```diff
@@ -48,8 +60,170 @@ def do_list(self, args):
...
+            if hammr_utils.is_uforge_exception(e):
+                return hammr_utils.handle_uforge_exception(e)
+            printer.out(str(e), printer.ERROR)
+            # return hammr_utils.handle_uforge_exception(e)
+
+    def help_launch(self):
+        doParser = self.arg_launch()
+        doParser.print_help()
+
+    def download_migration_binary(self):
```
Comment (segalaj) [inline hammr/commands/migration/migration.py]: this method contains duplicated code from scan.py, it should be extracted and reuse

File: hammr/utils/hammr_utils.py [hunk 7]
```diff
@@ -270,3 +270,29 @@ def extract_scan_id(image_uri):
         return int(match.group(1))
     else:
         return None
+
+def extract_appliance_id(image_uri):
```
Comment (segalaj) [inline hammr/utils/hammr_utils.py]: this seems to be unused

File: hammr/utils/hammr_utils.py [hunk 8]
```diff
@@ -270,3 +270,29 @@ def extract_scan_id(image_uri):
...
+    match = re.match( r'users/[^/]+/appliances/([0-9]+)($|/)', image_uri)
+    if match:
+        return int(match.group(1))
+    else:
+        return None
+
+def check_mandatory_migration(migration):
+    return migration
+
+def check_mandatory_migrate(migration):
```
Comment (segalaj) [inline hammr/utils/hammr_utils.py]: could be move to migration_utils isn't it?

File: hammr/utils/hammr_utils.py [hunk 9]
```diff
@@ -270,3 +270,29 @@ def extract_scan_id(image_uri):
...
+    if not "os" in migration:
+        printer.out("no attribute [os] for [migration]", printer.ERROR)
+        return
+    if not "source" in migration:
+        printer.out("no attribute [source] for [migration]", printer.ERROR)
+        return
+    if not "destination" in migration:
+        printer.out("no attribute [destination] for [migration]", printer.ERROR)
+        return
+    return migration
```
Comment (segalaj) [inline hammr/utils/hammr_utils.py]: the migration is never updated, why should we return it?

File: hammr/utils/hammr_utils.py [hunk 10]
```diff
@@ -270,3 +270,29 @@ def extract_scan_id(image_uri):
...
+    else:
+        return None
+
+def check_mandatory_migration(migration):
+    return migration
+
+def check_mandatory_migrate(migration):
+    if not "name" in migration:
+        printer.out("no attribute [name] for [migration]", printer.ERROR)
+        return
```
Comment (segalaj) [inline hammr/utils/hammr_utils.py]: shouldn't we raise instead?

File: hammr/utils/migration_utils.py [hunk 11]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
...
+
+
+def check_migration_configuration(args_file):
+    file = generics_utils.get_file(args_file)
+    if file is None:
+        raise Exception("No such file or directory: " + args_file)
```
Comment (segalaj) [inline hammr/utils/migration_utils.py]: I think, we should use more specific exception:
- create a MigrationException
OR
- use existing ones: TypeError, ValueError, etc...

File: hammr/utils/migration_utils.py [hunk 12]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
...
+        return
+    else:
+        for account in accounts:
+            if account.name == credAccountName:
+                if account.targetPlatform.name == targetPlatformName:
+                    return account
+        return
+
+
+def check_migration_configuration(args_file):
```
Comment (segalaj) [inline hammr/utils/migration_utils.py]: naming is misleading, it retrieves the migration configuration

File: hammr/utils/migration_utils.py [hunk 13]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
...
+        migrationConfig = hammr_utils.check_mandatory_migration(data["migration"])
+        migrationConfig = hammr_utils.check_mandatory_migrate(migrationConfig)
+        if migrationConfig is None:
+            raise Exception("check yours parameters in file: " + args_file)
+        return migrationConfig
+    else:
+        raise Exception("no migration section found")
+
+
+def check_target(api, login, migrationConfig):
```
Comment (segalaj) [inline hammr/utils/migration_utils.py]: naming is misleading, it retrieves the target format and target platform.

File: hammr/utils/migration_utils.py [hunk 14]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
...
+
+    targetPlatform = account_utils.get_target_platform_object(api, login,
+                                                              migrationConfig["destination"]["platform"])
+    if targetPlatform is None:
+        raise Exception("TargetPlatform type unknown: " + migrationConfig["destination"]["platform"])
+
+    return targetFormat, targetPlatform
+
+
+def check_publish_image(migrationConfig, targetFormat, file):
```
Comment (segalaj) [inline hammr/utils/migration_utils.py]: naming is misleading, it retrieves the publish image.

File: hammr/utils/migration_utils.py [hunk 15]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
...
+    if publishMethod:
+        myPublishImage = publishMethod(destination)
+        if myPublishImage is None:
+            raise Exception("check yours parameters in file: " + file)
+        return myPublishImage
+    else:
+        raise Exception("TargetFormat type is unsupported: " + targetFormat.format.name)
+
+
+def check_credAccount(api, login, migrationConfig, targetPlatform):
```
Comment (segalaj) [inline hammr/utils/migration_utils.py]: naming is misleading, it retrieves the account.

File: tests/integration/data/testMigrate.json [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+      "displayName": "weasel-jerome-vcenter2",
+      "esxHost": "esx4dev.hq.usharesoft.com",
+      "datastore": "esx4dev_data1_secure",
+      "network": "VM Network",
+      "account": {
+        "name": "account_name"
+      }
+    }
+  }
+}
```
Comment (segalaj) [inline tests/integration/data/testMigrate.json]: this files is stored in the resources of ITs, but used in UTs

File: tests/unit/commands/migration/test_migration.py [hunk 17]
```diff
@@ -71,6 +75,112 @@ def test_do_list_return_no_migration_message_when_there_is_no_migration(self, mo
...
+        # when
+        m.do_launch("--file config.json")
+
+        # then
+        mock_out.assert_called_with("file: config.json is validated", "OK")
+        self.assertEquals(mock_download.call_count, 1)
+
+    @patch("hammr.commands.migration.Migration.download_migration_binary")
+    @patch("ussclicore.utils.printer.out")
+    def test_do_launch_display_message_error_when_file_is_not_filled(self, mock_out, mock_download):
```
Comment (segalaj) [inline tests/unit/commands/migration/test_migration.py]: I'm not sure this test is relevant, this tests the ussclicore third party library

File: tests/unit/commands/migration/test_migration.py [hunk 18]
```diff
@@ -71,6 +75,112 @@ def test_do_list_return_no_migration_message_when_there_is_no_migration(self, mo
...
+        m.do_launch("--file")
+
+        # then
+        mock_out.assert_called_with("ERROR: In Arguments: argument --file: expected one argument", "ERROR")
+        self.assertEquals(mock_download.call_count, 0)
+
+    @patch("hammr.commands.migration.Migration.download_migration_binary")
+    @patch("ussclicore.utils.generics_utils.get_file")
+    @patch("ussclicore.utils.printer.out")
+    def test_do_launch_not_launch_download_migration_binary_when_file_is_not_found(self, mock_out, mock_get_file,
```
Comment (segalaj) [inline tests/unit/commands/migration/test_migration.py]: idem

File: tests/unit/commands/migration/test_migration.py [hunk 19]
```diff
@@ -71,6 +75,112 @@ def test_do_list_return_no_migration_message_when_there_is_no_migration(self, mo
...
+        m.do_launch("--file config.json")
+
+        # then
+        mock_out.assert_called_with("no migration section found", "ERROR")
+        self.assertEquals(mock_download.call_count, 0)
+
+    def test_createMigration_create_a_migration_with_all_stages(self):
+        # given
+        m = migration.Migration()
+        data = self.retrieve_jsonFile("tests/integration/data/testMigrate.json")
```
Comment (segalaj) [inline tests/unit/commands/migration/test_migration.py]: close to IT, maybe we could build the json

File: tests/unit/utils/test_migration_utils.py [hunk 20]
```diff
@@ -85,17 +85,135 @@ def test_migration_table_add_the_migration_cancelled_in_the_table(self, mock_tab
         self.assertEquals(mock_table_add_row.call_count, 1)
         mock_table_add_row.assert_any_call([1, "a migration", "Cancelled"])
 
+    def test_validateAccount_return_the_account_when_this_account_is_on_platform(self):
```
Comment (segalaj) [inline tests/unit/utils/test_migration_utils.py]: no more validateAccount

File: tests/unit/utils/test_migration_utils.py [hunk 21]
```diff
@@ -85,17 +85,135 @@ def test_migration_table_add_the_migration_cancelled_in_the_table(self, mock_tab
...
+
+        # when
+        accountRetrieved = migration_utils.check_account_from_platform(accountsOnPlatform.credAccounts.credAccount,
+                                                                       credAccountName,
+                                                                       targetPlatformName)
+
+        # then
+        self.assertEqual(accountRetrieved.name, account.name)
+
+    def test_validateAccount_return_the_account_when_this_account_is_on_platform1(self):
```
Comment (segalaj) [inline tests/unit/utils/test_migration_utils.py]: method name is not correct

File: tests/unit/utils/test_migration_utils.py [hunk 22]
```diff
@@ -85,17 +85,135 @@ def test_migration_table_add_the_migration_cancelled_in_the_table(self, mock_tab
...
+        # then
+        self.assertEqual(accountRetrieved, None)
+
+    @patch("ussclicore.utils.generics_utils.get_file")
+    def test_check_migration_configuration_raise_exception_if_no_file_is_retrieved(self, mock_get_file):
+        # given
+        args_file = "file_no_present.json"
+        mock_get_file.return_value = None
+
+        # when
```
Comment (segalaj) [inline tests/unit/utils/test_migration_utils.py]: opt: you could put:
    # when / then

File: tests/unit/utils/test_migration_utils.py [hunk 23]
```diff
@@ -85,17 +85,135 @@ def test_migration_table_add_the_migration_cancelled_in_the_table(self, mock_tab
...
+
+        # then
+        self.assertRaises(Exception, migration_utils.check_migration_configuration, args_file)
+
+    @patch("ussclicore.utils.generics_utils.get_file")
+    @patch("hammr.utils.hammr_utils.load_data")
+    @patch("hammr.utils.hammr_utils.check_mandatory_migration")
+    @patch("hammr.utils.hammr_utils.check_mandatory_migrate")
+    def test_check_migration_configuration_raise_exception_if_file_contain_migration_section_with_correct_parameters(
```
Comment (segalaj) [inline tests/unit/utils/test_migration_utils.py]: it does not raise exception

File: hammr/commands/migration/migration.py [hunk 24]
```diff
@@ -48,8 +60,170 @@ def do_list(self, args):
...
+
+        except ArgumentParserError as e:
+            printer.out("ERROR: In Arguments: " + str(e), printer.ERROR)
+            self.help_launch()
+
+        except Exception as e:
+            if hammr_utils.is_uforge_exception(e):
+                return hammr_utils.handle_uforge_exception(e)
+            printer.out(str(e), printer.ERROR)
+            # return hammr_utils.handle_uforge_exception(e)
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: done

File: hammr/utils/migration_utils.py [hunk 25]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
         table.add_row([migration.dbId, migration.name, status])
 
     return table
+
+
+def check_account_from_platform(accounts, credAccountName, targetPlatformName):
+    if accounts is None or len(accounts) == 0:
+        printer.out(
+            "No accounts available on the plateform.\n You can use the command 'hammr account create' to create an account.",
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: hammr/utils/migration_utils.py [hunk 26]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
...
+            printer.ERROR)
+        return
+    else:
+        for account in accounts:
+            if account.name == credAccountName:
+                if account.targetPlatform.name == targetPlatformName:
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: hammr/utils/migration_utils.py [hunk 27]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
         table.add_row([migration.dbId, migration.name, status])
 
     return table
+
+
+def check_account_from_platform(accounts, credAccountName, targetPlatformName):
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: I change to 'retrieve_account_from_platform'

File: hammr/utils/migration_utils.py [hunk 28]
```diff
@@ -29,3 +36,83 @@ def migration_table(migrations):
...
+    if accounts is None or len(accounts) == 0:
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: ok it's work

File: tests/unit/utils/test_migration_utils.py (hunk 22 above)
Comment (Drab-Hounet) [reply tests/unit/utils/test_migration_utils.py]: done

File: tests/unit/utils/test_migration_utils.py (hunk 23 above)
Comment (Drab-Hounet) [reply tests/unit/utils/test_migration_utils.py]: modified

File: hammr/utils/migration_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: hammr/utils/migration_utils.py (hunk 1 above)
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: hammr/utils/migration_utils.py (hunk 2 above)
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: hammr/commands/migration/migration.py (hunk 4 above)
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: done

File: hammr/commands/migration/migration.py (hunk 5 above)
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: ok I switch to Camel case

File: hammr/utils/hammr_utils.py (hunk 7 above)
Comment (Drab-Hounet) [reply hammr/utils/hammr_utils.py]: it's a mistake ! removed

File: hammr/utils/hammr_utils.py (hunk 8 above)
Comment (Drab-Hounet) [reply hammr/utils/hammr_utils.py]: I moved to migration_utils

File: hammr/utils/hammr_utils.py (hunk 7 above)
Comment (Drab-Hounet) [reply hammr/utils/hammr_utils.py]: done

File: hammr/utils/hammr_utils.py (hunk 9 above)
Comment (Drab-Hounet) [reply hammr/utils/hammr_utils.py]: ok I changed it

File: hammr/utils/hammr_utils.py (hunk 10 above)
Comment (Drab-Hounet) [reply hammr/utils/hammr_utils.py]: done

File: hammr/utils/migration_utils.py (hunk 12 above)
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: hammr/utils/migration_utils.py (hunk 13 above)
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: hammr/utils/migration_utils.py (hunk 15 above)
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: tests/integration/data/testMigrate.json (hunk 16 above)
Comment (Drab-Hounet) [reply tests/integration/data/testMigrate.json]: I deleted this file

File: tests/unit/commands/migration/test_migration.py (hunk 19 above)
Comment (Drab-Hounet) [reply tests/unit/commands/migration/test_migration.py]: done

File: hammr/utils/migration_utils.py (hunk 14 above)
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: hammr/commands/migration/migration.py (hunk 6 above)
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: extracted

**Comment by AliceAlbano** [comment]
Hello, I have a small comment about this PR (I know I am not in review but I received an email about it). In last hammr commits, we are trying to make an effort to respect python naming conventions (ie snake_case). I am well aware that the existing code base does not always respect this rule, but I think that new code should respect it. For instance, myMigrationConfig would become my_migration_config.
Do not hesitate to discuss it with me if you disagree ;)

File: hammr/commands/migration/migration.py [hunk 29]
```diff
@@ -48,8 +55,106 @@ def do_list(self, args):
...
+            doParser = self.arg_launch()
+
+            try:
+                doArgs = doParser.parse_args(shlex.split(args))
+            except SystemExit:
+                return 2
+
+            migrationConfig = migration_utils.retrieve_migration_configuration(doArgs.file)
+            targetFormat, targetPlatform = migration_utils.retrieve_targets(self.api, self.login, migrationConfig)
+            myPublishImage = migration_utils.retrieve_publish_image(migrationConfig, targetFormat, doArgs.file)
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: done

File: hammr/commands/migration/migration.py [hunk 30]
```diff
@@ -48,8 +55,106 @@ def do_list(self, args):
...
+            myCredAccount = migration_utils.retrieve_credAccount(self.api, self.login, migrationConfig, targetPlatform)
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: done, I also take account Alice's remark for the naming convention -> I changed myPublisImage to publish_image.

File: hammr/commands/migration/migration.py [hunk 31]
```diff
@@ -48,8 +55,106 @@ def do_list(self, args):
...
+
+            printer.out("file: " + doArgs.file + " is validated", printer.OK)
+
+            myMigration = self.createMigration(migrationConfig, targetFormat, myCredAccount, myPublishImage)
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: done

File: hammr/commands/migration/migration.py [hunk 32]
```diff
@@ -48,8 +55,106 @@ def do_list(self, args):
...
+
+            # call UForge API
+            self.api.Users(self.login).Migrations.Create(body=myMigration, element_name="ns1:migration")
+
+            # download migration binary
+            localUforgeMigrationPath = hammr_utils.download_binary(self.api, constants.MIGRATION_BINARY_NAME,
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: done

File: hammr/commands/migration/migration.py [hunk 33]
```diff
@@ -48,8 +55,106 @@ def do_list(self, args):
...
+        username = migrationConfig["source"]["user"]
+        name = migrationConfig["name"]
+        port = migrationConfig["source"]["ssh-port"]
+        password = migrationConfig["source"]["password"]
+        dir = "/tmp"
+        command = 'chmod +x ' + dir + '/' + constants.MIGRATION_BINARY_NAME + '; nohup ' + dir + '/' + constants.MIGRATION_BINARY_NAME + ' -U ' + uforgeURL + ' -u ' + uforgeLogin + ' -p ' + uforgePassword + ' -n \'' + name + '\'' + ' >/dev/null 2>&1 &'
+
+        summary = hammr_utils.deploy_and_launch_agent(constants.MIGRATION_BINARY_NAME, command, dir, hostname, username, port,
+                                            password, fileSrcPath)
+        for line in summary.split('\n'):
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: You're right!
done

File: hammr/commands/migration/migration.py [hunk 34]
```diff
@@ -48,8 +55,106 @@ def do_list(self, args):
...
+            print line
+        shutil.rmtree(constants.TMP_WORKING_DIR)
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: I replace it like it was before
done

File: hammr/commands/migration/migration.py [hunk 35]
```diff
@@ -48,8 +55,106 @@ def do_list(self, args):
...
+
+    def createMigration(self, migrationConfig, targetFormat, credAccount, publishImage):
+        myMigration = migration()
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/commands/migration/migration.py]: I change to the_migration

File: hammr/commands/scan/scan.py [hunk 36]
```diff
@@ -504,86 +483,33 @@ def help_delete(self):
...
+            overlay = "-o"
+
+        dir = "/tmp"
+        commandTest = 'chmod +x ' + dir + '/' + constants.SCAN_BINARY_NAME + ';' + dir + '/' \
+                      + constants.SCAN_BINARY_NAME + ' -u ' + uforgeLogin + ' -p ' + uforgePassword + ' -U ' \
+                      + uforgeURL + ' -P'
+
+        summary = deploy_and_launch_agent(constants.SCAN_BINARY_NAME, commandTest, dir, hostname, username, port,
+                                          password, fileSrcPath)
+        for line in summary.split('\n'):
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/commands/scan/scan.py]: done

File: hammr/commands/scan/scan.py [hunk 37]
```diff
@@ -504,86 +483,33 @@ def help_delete(self):
...
+            print line
+
+        command = 'chmod +x ' + dir + '/' + constants.SCAN_BINARY_NAME + '; nohup ' + dir + '/' \
+                  + constants.SCAN_BINARY_NAME + ' -u ' + uforgeLogin + ' -p ' + uforgePassword + ' -U ' + uforgeURL \
+                  + ' ' + overlay + ' -n \'' + args.name + '\' ' + exclude + ' >/dev/null 2>&1 &'
+
+        deploy_and_launch_agent(constants.SCAN_BINARY_NAME, command, dir, hostname, username, port, password,
+                                fileSrcPath)
+        shutil.rmtree(constants.TMP_WORKING_DIR)
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/commands/scan/scan.py]: remove to do_launch method

File: hammr/utils/hammr_utils.py [hunk 38]
```diff
@@ -235,38 +242,137 @@ def create_user_ssh_key(api, login, sshKey):
...
+    except Exception, e:
+        try:
+            transport.close()
+            client.close()
+        except:
+            pass
+        raise Exception("Caught exception: " + str(e))
+
+
+def launch_agent(command, dir, username, passwordSource, fileSrcPath, hostname, port, constantBinaryName):
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/hammr_utils.py]: ok done

File: hammr/utils/hammr_utils.py [hunk 39]
```diff
@@ -235,38 +242,137 @@ def create_user_ssh_key(api, login, sshKey):
...
+    try:
+        paramiko.util.load_host_keys(os.path.expanduser('~/.ssh/known_hosts'))
+    except IOError:
+        try:
+            paramiko.util.load_host_keys(os.path.expanduser('~/ssh/known_hosts'))
+        except IOError:
+            raise Exception("Unable to open host keys file")
+
+
+def download_binary_ssh(dir, username, password, fileSrcPath, hostname, port, constantBinaryName):
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/hammr_utils.py]: done

File: hammr/utils/migration_utils.py [hunk 40]
```diff
@@ -29,3 +36,90 @@ def migration_table(migrations):
...
+
+    return targetFormat, targetPlatform
+
+
+def retrieve_publish_image(migrationConfig, targetFormat, file):
+    destination = migrationConfig["destination"]
+    publishMethod = getattr(publish_utils,
+                            "publish_" + generics_utils.remove_special_chars(targetFormat.format.name), None)
+    if publishMethod:
+        myPublishImage = publishMethod(destination)
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: hammr/utils/migration_utils.py [hunk 41]
```diff
@@ -29,3 +36,90 @@ def migration_table(migrations):
...
+            raise Exception("check yours parameters in file: " + file)
+        return myPublishImage
+    else:
+        raise Exception("TargetFormat type is unsupported: " + targetFormat.format.name)
+
+
+def retrieve_credAccount(api, login, migrationConfig, targetPlatform):
+    accountConfig = migrationConfig["destination"]["account"]
+    accountName = accountConfig["name"]
+    # Get all cloud account on the plateform (for the user)
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

File: hammr/utils/migration_utils.py [hunk 42]
```diff
@@ -29,3 +36,90 @@ def migration_table(migrations):
...
+    accountsOnPlatform = api.Users(login).Accounts.Getall()
+    accountsOnPlatform = accountsOnPlatform.credAccounts.credAccount
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: I had to do this to retrieve the object of the response of the api.

File: hammr/utils/migration_utils.py [hunk 43]
```diff
@@ -29,3 +36,90 @@ def migration_table(migrations):
...
+
+    accountRetrieved = retrieve_account_from_platform(accountsOnPlatform, accountName,
+                                                      targetPlatform.name)
+    if accountRetrieved is None:
+        raise Exception("CredAccount unknown: " + accountName + " for targetPlatform : " + targetPlatform.name
+                        + "\n You can use the command 'hammr account create' to create an account.")
+
+    func = getattr(account_utils, generics_utils.remove_special_chars(targetPlatform.type), None)
```
[the remark this answers is not in the document]
Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: func is a method built in function of the targePlatform type
for exemple for Vcenter I will retrieve func = account_utils.vsphere()

File: hammr/utils/migration_utils.py [hunk 44]
```diff
@@ -29,3 +36,90 @@ def migration_table(migrations):
         table.add_row([migration.dbId, migration.name, status])
 
     return table
+
+
+def retrieve_account_from_platform(accounts, credAccountName, targetPlatformName):
```
Comment (lqueiroga) [inline hammr/utils/migration_utils.py]: naming: only retrieve_account it's ok

Comment (Drab-Hounet) [reply hammr/utils/migration_utils.py]: done

**Comment by lqueiroga** [comment]
Thanks for this PR. Unfortunately, we will not merge this one but rework it together.
See #361

**Pull request closed without merging**